src/backend/distributed/metadata/node_metadata.c

```diff
@@ -428,66 +428,141 @@ citus_disable_node(PG_FUNCTION_ARGS)
 {
 	text *nodeNameText = PG_GETARG_TEXT_P(0);
 	int32 nodePort = PG_GETARG_INT32(1);
+	bool forceDisableNode = PG_GETARG_BOOL(2);
+
 	char *nodeName = text_to_cstring(nodeNameText);
 	WorkerNode *workerNode = ModifiableWorkerNode(nodeName, nodePort);
-	bool isActive = false;
-	bool onlyConsiderActivePlacements = false;
-	MemoryContext savedContext = CurrentMemoryContext;
 
-	PG_TRY();
+	WorkerNode *firstWorkerNode = GetFirstPrimaryWorkerNode();
+	if (!forceDisableNode && firstWorkerNode &&
+		firstWorkerNode->nodeId == workerNode->nodeId)
 	{
-		if (NodeIsPrimary(workerNode))
-		{
-			/*
-			 * Delete reference table placements so they are not taken into account
-			 * for the check if there are placements after this.
-			 */
-			DeleteAllReferenceTablePlacementsFromNodeGroup(workerNode->groupId);
+		/*
+		 * We sync metadata async and optionally in the background worker,
+		 * it would mean that some nodes might get the updates while other
+		 * not. And, if the node metadata that is changing is the first
+		 * worker node, the problem gets nasty. We serialize modifications
+		 * to replicated tables by acquiring locks on the first worker node.
+		 *
+		 * If some nodes get the metadata changes and some do not, they'd be
+		 * acquiring the locks on different nodes. Hence, having the
+		 * possibility of diverged shard placements for the same shard.
+		 *
+		 * To prevent that, we currently do not allow disabling the first
+		 * worker node.
+		 */
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("disabling the first worker node in the "
+							   "metadata is not allowed"),
+						errhint("You can force disabling node, but this operation "
+								"might cause data loss: SELECT "
+								"citus_disable_node('%s', %d, force:=true);",
+								workerNode->workerName,
+								nodePort)));
+	}
 
-			if (NodeGroupHasShardPlacements(workerNode->groupId,
-											onlyConsiderActivePlacements))
-			{
-				ereport(NOTICE, (errmsg(
-									 "Node %s:%d has active shard placements. Some queries "
-									 "may fail after this operation. Use "
-									 "SELECT citus_activate_node('%s', %d) to activate this "
-									 "node back.",
-									 workerNode->workerName, nodePort,
-									 workerNode->workerName,
-									 nodePort)));
-			}
+	/*
+	 * First, locally mark the node as inactive. We'll later trigger background
+	 * worker to sync the metadata changes to the relevant nodes.
+	 */
+	bool isActive = false;
+	workerNode =
+		SetWorkerColumnLocalOnly(workerNode,
+								 Anum_pg_dist_node_isactive,
+								 BoolGetDatum(isActive));
+	if (NodeIsPrimary(workerNode))
```

Comment:
What should the behaviour be for non-primary nodes? (if there is a follower)

It kind of seems like it should cascade to secondaries.

Reply:
For secondaries, we update pg_dist_node.isactive = false, and let the bg daemon to update all the primaries with this information (hence implicitly updating the secondaries).

> It kind of seems like it should cascade to secondaries.

Do you have something else in mind regarding this?

Reply:
> Do you have something else in mind regarding this?

Update all secondaries with the same groupid at once, and disallow disabling secondaries.

It makes little sense for a primary to be disabled and a secondary to be enabled.

Reply:
> It makes little sense for a primary to be disabled and a secondary to be enabled.

I thought even if a primary is down, you might still want to continue querying the secondaries, say you are OK with the  last data you have on the replicas. If you explicitly disable secondaries, then you won't be able to query them as you'd get  errors like`ERROR:  no active placements were found for shard 102650`.

Probably your suggestion makes more sense. But, it triggered some changes in activate_node as well. Once we agree the behavior, I'll add some tests.

Reply:
alright, if disabling them separately keeps things simpler that might be ok, since this is going to be a quite unusual combination

Reply:
reverted those commits for now